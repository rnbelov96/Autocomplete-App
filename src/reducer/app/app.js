import countries from '../../data/countries';

const initialState = {
  countries,
  suggestions: [],
  inputValue: '',
};

const ActionTypes = {
  CHANGE_INPUT: 'CHANGE_INPUT',
  FIND_MATCH: 'FIND_MATCH',
};

const ActionCreators = {
  changeInput: data => ({ type: ActionTypes.CHANGE_INPUT, payload: data }),

  findMatch: () => ({ type: ActionTypes.FIND_MATCH }),
};

const reducer = (state = initialState, action) => {
  switch (action.type) {
    case ActionTypes.CHANGE_INPUT:
      return {
        ...state,
        inputValue: action.payload,
      };

    case ActionTypes.FIND_MATCH:
      return state.inputValue === ''
        ? { ...state, suggestions: [] }
        : {
          ...state,
          suggestions: state.countries.filter(el => el
            .toLowerCase()
            .startsWith(state.inputValue.toLowerCase())),
        };
    default:
      return state;
  }
};

export {
  reducer as appReducer,
  ActionCreators as AppActionCreators,
  ActionTypes as AppActionTypes,
};
