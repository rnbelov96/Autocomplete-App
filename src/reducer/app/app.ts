import {
  AppActionConstType,
  AppActionType,
  AppInitialStateType,
  ChangeInputActionType,
  FindMatchActionType,
} from '@/types/redux/app-reducer';
import countries from '../../data/countries';

const initialState: AppInitialStateType = {
  countries,
  suggestions: [],
  inputValue: '',
};

const ActionTypes: AppActionConstType = {
  CHANGE_INPUT: 'CHANGE_INPUT',
  FIND_MATCH: 'FIND_MATCH',
};

const ActionCreators = {
  changeInput: (data: string): ChangeInputActionType => ({
    type: ActionTypes.CHANGE_INPUT,
    payload: data,
  }),

  findMatch: (): FindMatchActionType => ({ type: ActionTypes.FIND_MATCH }),
};

const reducer = (
  state: AppInitialStateType = initialState,
  action: AppActionType,
) => {
  switch (action.type) {
    case ActionTypes.CHANGE_INPUT:
      return {
        ...state,
        inputValue: action.payload,
      };

    case ActionTypes.FIND_MATCH:
      return state.inputValue === ''
        ? { ...state, suggestions: [] }
        : {
          ...state,
          suggestions: state.countries.filter(el => el.toLowerCase().startsWith(state.inputValue.toLowerCase())),
        };
    default:
      return state;
  }
};

export {
  reducer as appReducer,
  ActionCreators as AppActionCreators,
  ActionTypes as AppActionTypes,
};
