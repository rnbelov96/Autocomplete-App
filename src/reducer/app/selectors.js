export const getSuggestions = state => state.app.suggestions;

export const getInputValue = state => state.app.inputValue;
