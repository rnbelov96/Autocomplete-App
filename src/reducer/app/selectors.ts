import { FullStateType } from '@/types/general-types';

export const getSuggestions = (state: FullStateType) => state.app.suggestions;

export const getInputValue = (state: FullStateType) => state.app.inputValue;
