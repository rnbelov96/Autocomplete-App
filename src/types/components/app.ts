export type AppPropsType = {
  onInputChange: (data: string) => void,
  inputValue: string,
  suggestions: string[]
};
