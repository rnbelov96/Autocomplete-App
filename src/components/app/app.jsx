import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { getInputValue, getSuggestions } from '@/reducer/app/selectors';
import { AppActionCreators } from '@/reducer/app/app';

const AutoComplete = ({ inputValue, onChange, suggestions }) => (
  <div className="container">
    <div className="auto-complete">
      <h1 className="auto-complete__title">Auto Complete</h1>
      <input
        className="auto-complete__input"
        type="text"
        value={inputValue}
        onChange={e => {
          onChange(e.target.value);
        }}
      />
      <ul
        style={suggestions.length === 0 ? { display: 'none' } : {}}
        className="auto-complete__suggestions"
      >
        {suggestions.map(el => (
          <li key={el} className="auto-complete__suggestions-item">
            {el}
          </li>
        ))}
      </ul>
    </div>
    <div className="auto-complete__suggestions-count">{`Suggestions: ${suggestions.length}`}</div>
  </div>
);

AutoComplete.propTypes = {
  inputValue: PropTypes.string.isRequired,
  onChange: PropTypes.func.isRequired,
  suggestions: PropTypes.arrayOf(PropTypes.string).isRequired,
};

const mapStateToProps = state => ({
  suggestions: getSuggestions(state),
  inputValue: getInputValue(state),
});

const mapDispatchToProps = dispatch => ({
  onChange: data => {
    dispatch(AppActionCreators.changeInput(data));
    dispatch(AppActionCreators.findMatch());
  },
});

export default connect(mapStateToProps, mapDispatchToProps)(AutoComplete);
