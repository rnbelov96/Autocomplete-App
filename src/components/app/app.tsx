import React, { Dispatch } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { getInputValue, getSuggestions } from '@/reducer/app/selectors';
import { AppActionCreators } from '@/reducer/app/app';
import { AppPropsType } from '@/types/components/app';
import { FullStateType } from '@/types/general-types';
import { AppActionType } from '@/types/redux/app-reducer';

const AutoComplete: React.FunctionComponent<AppPropsType> = ({ inputValue, onInputChange, suggestions }) => (
  <div className="container">
    <div className="auto-complete">
      <h1 className="auto-complete__title">Auto Complete</h1>
      <input
        className="auto-complete__input"
        type="text"
        value={inputValue}
        onChange={e => {
          onInputChange(e.target.value);
        }}
      />
      <ul
        style={suggestions.length === 0 ? { display: 'none' } : {}}
        className="auto-complete__suggestions"
      >
        {suggestions.map(el => (
          <li key={el} className="auto-complete__suggestions-item">
            {el}
          </li>
        ))}
      </ul>
    </div>
    <div className="auto-complete__suggestions-count">{`Suggestions: ${suggestions.length}`}</div>
  </div>
);

const mapStateToProps = (state: FullStateType) => ({
  suggestions: getSuggestions(state),
  inputValue: getInputValue(state),
});

const mapDispatchToProps = (dispatch: Dispatch<AppActionType>) => ({
  onInputChange: (data: string) => {
    dispatch(AppActionCreators.changeInput(data));
    dispatch(AppActionCreators.findMatch());
  },
});

export default connect(mapStateToProps, mapDispatchToProps)(AutoComplete);
