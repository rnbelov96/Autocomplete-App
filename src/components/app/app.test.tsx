import React from 'react';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import { App } from './app';

const mockStore = configureStore([]);

describe('App', () => {
  test('should render initial screen', () => {
    const store = mockStore({
      app: {
        countries: ['Russian', 'USA'],
        inputValue: '',
        suggestions: [],
      },
    });
    const tree = renderer
      .create(
        <Provider store={store}>
          <App />
        </Provider>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('should render screen with results', () => {
    const store = mockStore({
      app: {
        countries: ['Russian', 'USA', 'Romania'],
        inputValue: 'R',
        suggestions: ['Russian', 'Romania'],
      },
    });
    const tree = renderer
      .create(
        <Provider store={store}>
          <App />
        </Provider>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
