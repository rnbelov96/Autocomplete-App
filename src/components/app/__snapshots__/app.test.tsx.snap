// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should render initial screen 1`] = `
<div
  className="container"
>
  <div
    className="auto-complete"
  >
    <h1
      className="auto-complete__title"
    >
      Auto Complete
    </h1>
    <input
      className="auto-complete__input"
      onChange={[Function]}
      type="text"
      value=""
    />
    <ul
      className="auto-complete__suggestions"
      style={
        Object {
          "display": "none",
        }
      }
    />
  </div>
  <div
    className="auto-complete__suggestions-count"
  >
    Suggestions: 0
  </div>
</div>
`;

exports[`App should render screen with results 1`] = `
<div
  className="container"
>
  <div
    className="auto-complete"
  >
    <h1
      className="auto-complete__title"
    >
      Auto Complete
    </h1>
    <input
      className="auto-complete__input"
      onChange={[Function]}
      type="text"
      value="R"
    />
    <ul
      className="auto-complete__suggestions"
      style={Object {}}
    >
      <li
        className="auto-complete__suggestions-item"
      >
        Russian
      </li>
      <li
        className="auto-complete__suggestions-item"
      >
        Romania
      </li>
    </ul>
  </div>
  <div
    className="auto-complete__suggestions-count"
  >
    Suggestions: 2
  </div>
</div>
`;
